lec06-pop-models.Rmd

```diff
@@ -2,3 +2,360 @@
+```
+We can also tell whether the fixed points are stable or unstable by looking at the sign of $dN/dt$ 
+on either side of the fixed point. For a one-dimensional system, our phase portrait is a line of $N$. 
+We mark the fixed point(s) with circles, and we draw an arrow to the left wherever $dN/dt$ is negative, 
+and an arrow to the right wherever $dN/dt$ is positive in the regions between fixed points. 
+
+<!-- diagram of phase portrait --> 
+
+This model is a bit boring - there's only one fixed point, and it's unstable. This makes sense - 
+we are talking about exponential growth, after all. There's no upper limit. 
+
+--------
+
+Of course in real life there must be an upper limit to growth, otherwise the surface of the earth would be
+covered in bacteria several feet thick in a matter of hours. Let's do a quick calculation for fun - how long
+would it take for a single bacterial cell undergoing exponential growth to take over the world?
+
+```{r}
+cell_volume = 0.6 #micrometres cubed - E. coli cell volume is ~0.6 to 0.7 micrometres^3
+growth_rate = 3 # let's say cells divide three times per hour - every 20 minutes (max growth rate of E. coli in good food conditions)
+
+N_t <- function(N0,r,t) { # r is growth rate - divisions per hour, t is time in hours, N0 is starting population
+  return(N0*2^(r*t-1)) # the -1 is because the first doubling is included with the factor of 2
+}
+
+radius_earth = 6000 # approximately 6000 km 
+SA_earth = 4*pi*radius_earth^2 # surface area of earth in km^2
+
+surface_covered <- function(N0,r,t) { # calculate the volume taken up by bacteria in m on the earth's surface
+  bac_volume <- N_t(N0,growth_rate,t)*cell_volume # in micrometres cubed
+  height_km = bac_volume*(1e-27)/SA_earth # assume the earth is flat compared to the height of our surface - volume is surface area x height of surface above earth. This is height in km
+  #surface_volume = height*SA_earth
+  
+  height_m = height_km*1000 # 1000 m in km
+  
+  return(height_m)
+}
+
+t<-37
+message("The bacteria would be ", round(surface_covered(2,growth_rate,t),1),  " metres thick on the surface of the earth after ", t, " hours.") #yikes. In less than two days, we would be submerged in bacteria. 
+```
+
+Let's look at another model, the *logistic growth model*, which most of you will have seen before:
+
+$$\frac{dN}{dt} = rN\left(1-\frac{N}{K}\right)$$
+
+$N$ is the number of organisms, $r$ is the growth rate, and $K$ is the carrying capacity of the system. 
+
+Notice that when $N$ is much smaller than $K$, this reduces to the exponential growth model we just looked at since $N/K \approx 0$. 
+
+Let's again plot $dN/dt$ vs. $N$ to find the fixed points:
+
+```{r}
+N_seq <- seq(-20,100) # make a sequence of N values to plot dN/dt against
+
+logistic_eqn <- function(N,r,K) { # calculate dN/dt for the logistic equation
+  return(r*N*(1-N/K))
+}
+
+dN_dt = logistic_eqn(N_seq,r=0.5,K=80)  # notice the vectorized implementation
```

Comment:
I only had time to touch briefly on for loops during the lecture in class, so maybe just clarify that vectorized here refers to that R performs matrix algebra faster under the hood, than if we would loop through each element manually within a for loop. And the code also become easier to read with the current notation.